subsys/task_wdt/task_wdt.c

```diff
@@ -0,0 +1,212 @@
+#include <power/reboot.h>
+#include <device.h>
+#include <errno.h>
+
+#define LOG_LEVEL CONFIG_WDT_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(task_wdt);
+
+/*
+ * This dummy channel is used to continue feeding the hardware watchdog if the
+ * task watchdog timeouts are too long for regular updates
+ */
+#define TASK_WDT_BACKGROUND_CHANNEL (-1)
+
+/*
+ * Task watchdog channel data
+ */
+struct task_wdt_channel {
+	/* period in milliseconds used to reset the timeout, set to 0 to
+	 * indicate that the channel is available
+	 */
+	uint32_t reload_period;
+	/* abs. ticks when this channel expires (updated by task_wdt_feed) */
+	int64_t timeout_abs_ticks;
+	/* user data passed to the callback function */
+	void *user_data;
+	/* function to be called when watchdog timer expired */
+	task_wdt_callback_t callback;
+};
+
+/* array of all task watchdog channels */
+static struct task_wdt_channel channels[CONFIG_TASK_WDT_CHANNELS];
+
+/* timer used for watchdog handling */
+static struct k_timer timer;
+
+#ifdef CONFIG_TASK_WDT_HW_FALLBACK
+/* pointer to the hardware watchdog used as a fallback */
+static const struct device *hw_wdt_dev;
+static int hw_wdt_channel;
+static bool hw_wdt_started;
+#endif
+
+/**
+ * @brief Task watchdog timer callback.
+ *
+ * If the device operates as intended, this function will never be called,
+ * as the timer is continuously restarted with the next due timeout in the
+ * task_wdt_feed() function.
+ *
+ * If all task watchdogs have longer timeouts than the hardware watchdog,
+ * this function is called regularly (via the background channel). This
+ * should be avoided by setting CONFIG_TASK_WDT_MIN_TIMEOUT to the minimum
+ * task watchdog timeout used in the application.
+ *
+ * @param timer_id Pointer to the timer which called the function
+ */
+static void task_wdt_trigger(struct k_timer *timer_id)
+{
+	int channel_id = (int)k_timer_user_data_get(timer_id);
+
+#ifdef CONFIG_TASK_WDT_HW_FALLBACK
+	if (channel_id == TASK_WDT_BACKGROUND_CHANNEL) {
+		if (hw_wdt_dev) {
+			wdt_feed(hw_wdt_dev, 0);
+		}
+		return;
+	}
+#endif
+
+	if (channels[channel_id].reload_period == 0) {
+		/* channel was deleted */
+		return;
+	} else if (channels[channel_id].callback) {
+		channels[channel_id].callback(channel_id,
+			channels[channel_id].user_data);
+	} else {
+		sys_reboot(SYS_REBOOT_COLD);
+	}
+}
+
+int task_wdt_init(const struct device *hw_wdt)
+{
+	if (hw_wdt) {
+#ifdef CONFIG_TASK_WDT_HW_FALLBACK
+		struct wdt_timeout_cfg wdt_config;
+
+		wdt_config.flags = WDT_FLAG_RESET_SOC;
+		wdt_config.window.min = 0U;
+		wdt_config.window.max = CONFIG_TASK_WDT_MIN_TIMEOUT +
+			CONFIG_TASK_WDT_HW_FALLBACK_DELAY;
+		wdt_config.callback = NULL;
+
+		hw_wdt_dev = hw_wdt;
+		hw_wdt_channel = wdt_install_timeout(hw_wdt_dev, &wdt_config);
+#else
+		return -ENOTSUP;
+#endif
+	}
+
+	k_timer_init(&timer, task_wdt_trigger, NULL);
+
+	return 0;
+}
+
+int task_wdt_add(uint32_t reload_period, task_wdt_callback_t callback,
```

Comment:
Would be neat to leverage this infrastructure for tasks created internally by zephyr as well. (i.e sysworkq, idle, shell task) but I don't see any easy way to do that. I guess in a future PR Kconfig options could be added to those task creation blocks to enable task watchdogs or do you see a better way @martinjaeger ?

Reply:
Interesting question. Adding task watchdogs by default is certainly not an option, as it does add some overhead and not everyone might want it. So we'd probably need a dedicated Kconfig option per internal subsystem to enable/disable the `task_wdt` supervision.

I was also thinking about whether the `task_wdt` should better become a kernel service, so it would be more generally available. And would it then make sense to integrate the hardware watchdog fallback directly into the scheduler instead of using it to monitor indirectly if `k_timer` callbacks are still happening?